### PR TITLE
FEATURE: Fit SideTitle inside axis bounding box (prevent title overflow on edge)

### LINKED ISSUE #603: bottomTitles - Last title overflowed
**Describe the bug**
The problem is, that the last title on the x-axis on SideTitle overflowed. Is there a way that the last title don't overflowed?

**Screenshot**
Bildschirmfoto 2021-03-15 um 14 28 25

**Versions**
 Flutter: (Channel master, 2.1.0-13.0.pre.64)
fl_chart:  ^0.20.1

**To Reproduce**
```dart
import 'package:flutter/foundation.dart';
import 'package:flutter/material.dart';
import 'package:fl_chart/fl_chart.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      showPerformanceOverlay: false,
      theme: ThemeData(
        primaryColor: const Color(0xff262545),
        primaryColorDark: const Color(0xff201f39),
        brightness: Brightness.dark,
      ),
      home: const MyHomePage(title: 'fl_chart'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({Key key, this.title}) : super(key: key);
  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  final _controller = PageController(initialPage: 0);
  final _pages = [
    LineChartPage(),
  ];

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: SafeArea(
        child: PageView(
          physics: kIsWeb
              ? NeverScrollableScrollPhysics()
              : AlwaysScrollableScrollPhysics(),
          controller: _controller,
          children: _pages,
        ),
      ),
    );
  }
}

class LineChartPage extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Container(
      color: const Color(0xff262545),
      child: ListView(
        children: <Widget>[
          const Align(
            alignment: Alignment.centerLeft,
            child: Padding(
              padding: EdgeInsets.only(
                left: 36.0,
                top: 24,
              ),
            ),
          ),
          const SizedBox(
            height: 8,
          ),
          Padding(
            padding: const EdgeInsets.only(
              left: 28,
              right: 28,
            ),
            child: LineChartSample1(),
          ),
        ],
      ),
    );
  }
}

class LineChartSample1 extends StatefulWidget {
  @override
  State<StatefulWidget> createState() => LineChartSample1State();
}

class LineChartSample1State extends State<LineChartSample1> {
  bool isShowingMainData;

  @override
  void initState() {
    super.initState();
    isShowingMainData = true;
  }

  @override
  Widget build(BuildContext context) {
    return Container(
      height: MediaQuery.of(context).size.height * 0.9,
      decoration: const BoxDecoration(
        borderRadius: BorderRadius.all(Radius.circular(18)),
        gradient: LinearGradient(
          colors: [
            Color(0xff2c274c),
            Color(0xff46426c),
          ],
          begin: Alignment.bottomCenter,
          end: Alignment.topCenter,
        ),
      ),
      child: Stack(
        children: <Widget>[
          Column(
            crossAxisAlignment: CrossAxisAlignment.stretch,
            children: <Widget>[
              const SizedBox(
                height: 37,
              ),
              const SizedBox(
                height: 37,
              ),
              Expanded(
                child: Padding(
                  padding: const EdgeInsets.only(right: 16.0, left: 6.0),
                  child: LineChart(
                    sampleData1(),
                  ),
                ),
              ),
              const SizedBox(
                height: 10,
              ),
            ],
          ),
        ],
      ),
    );
  }
}

LineChartData sampleData1() {
  return LineChartData(
    lineTouchData: LineTouchData(
      touchTooltipData: LineTouchTooltipData(
        tooltipBgColor: Colors.blueGrey.withOpacity(0.8),
      ),
      touchCallback: (LineTouchResponse touchResponse) {},
      handleBuiltInTouches: true,
    ),
    gridData: FlGridData(
      show: false,
    ),
    titlesData: FlTitlesData(
      bottomTitles: SideTitles(
        showTitles: true,
        reservedSize: 22,
        getTextStyles: (value) => const TextStyle(
          color: Color(0xff72719b),
          fontWeight: FontWeight.bold,
          fontSize: 16,
        ),
        margin: 10,
        getTitles: (value) {
          switch (value.toInt()) {
            case 1:
              return '30 September 2020';
            case 7:
              return '30 Oktocber 2020';
            case 13:
              return '30 December 2020';
          }
          return '';
        },
      ),
      leftTitles: SideTitles(
        showTitles: true,
        getTextStyles: (value) => const TextStyle(
          color: Color(0xff75729e),
          fontWeight: FontWeight.bold,
          fontSize: 14,
        ),
        getTitles: (value) {
          switch (value.toInt()) {
            case 1:
              return '1m';
            case 2:
              return '2m';
            case 3:
              return '3m';
            case 4:
              return '5m';
          }
          return '';
        },
        margin: 8,
        reservedSize: 30,
      ),
    ),
    borderData: FlBorderData(
      show: true,
      border: const Border(
        bottom: BorderSide(
          color: Color(0xff4e4965),
          width: 4,
        ),
        left: BorderSide(
          color: Colors.transparent,
        ),
        right: BorderSide(
          color: Colors.transparent,
        ),
        top: BorderSide(
          color: Colors.transparent,
        ),
      ),
    ),
    minX: 0,
    maxX: 13,
    maxY: 4,
    minY: 0,
    lineBarsData: linesBarData1(),
  );
}

List<LineChartBarData> linesBarData1() {
  final LineChartBarData lineChartBarData1 = LineChartBarData(
    spots: [
      FlSpot(1, 1),
      FlSpot(3, 1.5),
      FlSpot(5, 1.4),
      FlSpot(7, 3.4),
      FlSpot(10, 2),
      FlSpot(12, 2.2),
      FlSpot(13, 1.8),
    ],
    isCurved: true,
    colors: [
      const Color(0xff4af699),
    ],
    barWidth: 8,
    isStrokeCapRound: true,
    dotData: FlDotData(
      show: false,
    ),
    belowBarData: BarAreaData(
      show: false,
    ),
  );
  return [
    lineChartBarData1,
  ];
}
```

### PR DESCRIPTION
**Pull request by dhiyaaulauliyaa**
Hi All!

I'm adding a feature to add an option for SideTitle to be fitted inside its corresponding axis bounding box.
This will fix #603. Below is the illustration:
- Before Using FitInside (SideTitle is overflowed): 
Screenshot_1675854250
- After Using FitInside (SideTitle is translated, hence not overflowed):
Screenshot_1675854231

What is added:
- FitInside data class (SideTitleFitInsideData) in lib/src/chart/base/axis_chart/axis_chart_data.dart
- FitInside parameter on SideTitleWidget
- Test for data class (SideTitleFitInsideData)
- Sample usage in example/lib/presentation/samples/line/line_chart_sample3.dart

### CODE REVIEW

**Comment by imaNNeo** [comment]
Thank you for this great feature!
First of all, please check the below bug:
When I hover my mouse, it hides the titles. 
https://user-images.githubusercontent.com/7009300/217529221-63d57390-f75c-4799-96e1-7269a8d61c93.mov

It seems it is related to this code:
```dart
ValueListenableBuilder(
  valueListenable: _childSize,
  builder: (_, double? size, Widget? child) {
    return Transform.translate(
      offset: _getOffset(),
      child: Opacity(
        opacity: fitInside.enabled && size == null ? 0 : 1,
        child: child,
      ),
    );
  },
)
```

Also, there are some failing CI workflows. Please check and fix those issues.
Don't forget to update the documentation. Also, you should write a changelog for it.
So it would be great if you read our [contributing guidelines](https://github.com/imaNNeo/fl_chart/blob/master/CONTRIBUTING.md).

**Comment by dhiyaaulauliyaa** [comment]
> Thank you for this great feature! First of all, please check the below bug: When I hover my mouse, it hides the titles.
> 
>  hover-bug.mov 
> It seems it is related to this code:
> 
> ```dart
> ValueListenableBuilder(
>   valueListenable: _childSize,
>   builder: (_, double? size, Widget? child) {
>     return Transform.translate(
>       offset: _getOffset(),
>       child: Opacity(
>         opacity: fitInside.enabled && size == null ? 0 : 1,
>         child: child,
>       ),
>     );
>   },
> )
> ```
> 
> Also, there are some failing CI workflows. Please check and fix those issues. Don't forget to update the documentation. Also, you should write a changelog for it. So it would be great if you read our [contributing guidelines](https://github.com/imaNNeo/fl_chart/blob/master/CONTRIBUTING.md).

Thanks for the response.

The bug has been fixed. Now the title won't blink when the chart is on hover.
Also, I have added the widget test and increased the coverage of my added changes.
For the docs, I have updated the documentation and added a changelog about this added feature.

File: lib/src/chart/base/axis_chart/axis_chart_helper.dart [hunk 1]
```diff
@@ -51,4 +53,52 @@ class AxisChartHelper {
...
+    required double? childSize,
+    required double parentAxisSize,
+    required double axisPosition,
+    required double distanceFromEdge,
+  }) {
+    if (childSize == null) return Offset.zero;
+
+    // Find title alignment along its axis
+    final axisMid = parentAxisSize / 2;
+    final mainAxisAligment = (axisPosition - axisMid).isNegative
```
Comment (imaNNeo) [inline lib/src/chart/base/axis_chart/axis_chart_helper.dart:72]: There is a typo issue in `mainAxisAligment`. Please fix it.

**Comment by imaNNeo** [comment]
Hi, we are ready to merge after fixing the typo. 

Thanks!

File: lib/src/chart/base/axis_chart/axis_chart_helper.dart (hunk 1 above)
Comment (dhiyaaulauliyaa) [reply lib/src/chart/base/axis_chart/axis_chart_helper.dart]: I have fix the typo in the latest commit

**Comment by imaNNeo** [comment]
Thank you very much.

**Comment by HornyPony** [comment]
Hi! How can I usee this option?

**Pull request merged**